pyodide_build/recipe/skeleton.py

```diff
@@ -82,6 +82,68 @@ def _find_wheel(pypi_metadata: MetadataDict, native: bool = False) -> URLDict |
     return None
 
 
+def _make_predictable_url(
+    package: str, version: str, source_type: Literal["wheel", "sdist"], filename: str
+) -> str | None:
+    """
+    Create a predictable URL for a PyPI package based on PyPI's conventions,
+    documented in https://docs.pypi.org/api/#predictable-urls.
+
+    Parameters
+    ----------
+    package
+        The package name
+    version
+        The package version
+    source_type
+        Either "wheel" or "sdist"
+    filename
+        The full filename (used for wheels to extract required tags)
+
+    Returns
+    -------
+    A predictable URL for the package file, or None if the URL could not be
+    constructed.
+    """
+    from packaging import utils as packaging_utils
+
+    host = "https://files.pythonhosted.org"
+
+    # convert hyphens to underscores for the package name, as
+    # PyPI uses underscores in the URL and packaging does not
+    # handle this.
+    package_name = packaging_utils.canonicalize_name(package)
+    package_url_name = package_name.replace("-", "_")
+
+    if source_type == "sdist":
+        return f"{host}/packages/source/{package_name[0]}/{package_url_name}/{package_url_name}-{version}.tar.gz"
+
+    elif source_type == "wheel":
+        try:
+            # extract Python tag
+            parts = filename.split("-")
```

Comment:
Probably you can use [parse_wheel_filename](https://packaging.pypa.io/en/stable/utils.html#packaging.utils.parse_wheel_filename) here, and use the tag part.

Usages in pyodide-build:

https://github.com/pyodide/pyodide-build/blob/005cabbce3c3f2a09210242ca2c934ec1cad2b07/pyodide_build/common.py#L58-L59